lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java

```diff
@@ -3775,4 +3777,120 @@ public void testRefreshAndRollbackConcurrently() throws Exception {
       IOUtils.close(sm, dir);
     }
   }
+
+  public void testRandomOperations() throws Exception {
+    Directory dir = newDirectory();
+    IndexWriterConfig iwc = newIndexWriterConfig();
+    iwc.setMergePolicy(new FilterMergePolicy(newMergePolicy()) {
+      boolean keepFullyDeletedSegment = random().nextBoolean();
+      @Override
+      public boolean keepFullyDeletedSegment(IOSupplier<CodecReader> readerIOSupplier) {
+        return keepFullyDeletedSegment;
+      }
+    });
+    IndexWriter writer = new IndexWriter(dir, iwc);
+    SearcherManager sm = new SearcherManager(writer, new SearcherFactory());
+    Semaphore permits = new Semaphore(100 + random().nextInt(1000));
+    boolean singleDoc = random().nextBoolean();
+    Thread[] threads = new Thread[1 + random().nextInt(4)];
+    CountDownLatch latch = new CountDownLatch(threads.length);
+    for (int i = 0; i < threads.length; i++) {
+      threads[i] = new Thread(() -> {
+        latch.countDown();
+        try {
+          latch.await();
+          while (permits.tryAcquire()) {
+            String id = singleDoc ? "1" : Integer.toString(random().nextInt(10));
+            Document doc = new Document();
+            doc.add(new StringField("id", id, Field.Store.YES));
+            if (random().nextInt(10) <= 2) {
+              writer.updateDocument(new Term("id", id), doc);
+            } else if (random().nextInt(10) <= 2) {
+              writer.deleteDocuments(new Term("id", id));
+            } else {
+              writer.addDocument(doc);
+            }
+            if (random().nextInt(100) < 10) {
+              sm.maybeRefreshBlocking();
+            }
+            if (random().nextInt(100) < 5) {
+              writer.commit();
+            }
+            if (random().nextInt(100) < 1) {
+              writer.forceMerge(1 + random().nextInt(10), random().nextBoolean());
+            }
+          }
+        } catch (Exception e) {
+          throw new AssertionError(e);
+        }
+      });
+      threads[i].start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    sm.maybeRefreshBlocking();
+    IOUtils.close(writer, sm, dir);
+  }
+
+  public void testRandomOperationsWithSoftDeletes() throws Exception {
+    Directory dir = newDirectory();
```

Comment:
can we use try-with-resources to ensure proper closure of Directory, IndexWriter, and SearcherManager on failure?

Reply:
Hmm I don't see where to officially +1 this in github, but thanks for sharing these tests!